### ISSUE TITLE
Client based rate limit

### ISSUE THREAD

**Issue by juhrmacher** [elasticio/rest-api-component#129]
As an integrator I would like to set a rate limit on the REST calls that my integration is executing on the external systems so that I could make sure my integration won't overload external systems and comply with API rate limit of external system

**Comment by shulkaolka** [comment]
In sprint 42 investigation is done. Some concept is here https://github.com/elasticio/rest-api-component/pull/131
It is used Maester as storage for information about usage and time.

**Comment by shulkaolka** [comment]
The open questions are:
1. How to connect limit to proper credentials
2. In what way handle with 429 code (try after some timeout, or only throw an error)
3. How to configure limit's options (metadata, config fields, envs)

**Comment by kirill-levitskiy** [comment]
1. We don't need to provide a limits configuration on a client's side. That's why we don't need to link limits to credentials.
2. Default behavior should include `rebound` feature to reprocess a message in case of limit exceeded. Additionally, we can add switcher for additional behavior - to drop a message in this case.
3. A client can't manage server's limits and control them. That's we we should rely on limits state that returned from a server's side.

Additional information:
- [Standard behavior for REST calls](https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html#ratelimit-headers-currently-used-on-the-web)
- [Salesforce implementation](https://developer.salesforce.com/docs/atlas.en-us.salesforce_app_limits_cheatsheet.meta/salesforce_app_limits_cheatsheet/salesforce_app_limits_platform_api.htm)
- [Microsoft Dynamics CRM implementation](https://docs.microsoft.com/en-us/powerapps/developer/common-data-service/api-limits#http-response-headers)
- [Docusign implementation](https://developers.docusign.com/esign-soap-api/reference/introduction-changes/api-call-limits#:~:text=DocuSign%20has%20implemented%20the%20following,calls%20per%20hour%20per%20account.)

**Comment by zubairov** [comment]
I totally agree with @shulkaolka that a proper rate limit should be 

(a) centrally managed so that even multiple integrations running in parallel should respect it

and 

(b) limits should be associated with credentials so that multiple flows and components of the same flow that uses same credential would respect the limits

and if combine above with 

(c) reaction on some HTTP Response and/or combination of HTTP Response Headers (mentioned by Kirill), more on that below

Now taking about complexity - my experience in distributed programming tells me it's close to impossible to implement a 100% reliable client-based rate control on top of number of distributed clients. We could only do that with a complex algorithm of allocating budgets to number of clients and coordinating between them dynamically. 

Nevertheless this feature would be a great functionality compared to what we have now - namely no rate limit at all.

However, even a rudimentary locally-controlled rate control could be still a good option taking into account simplicity of implementation. 

So for example when we do following:

(d) For the local rate control we only need to locally rate limit on HTTP calls without need of central coordination. 

(e) Component will have a combo-box or text field where a max number of requests per second can be configured, e.g. `0.1` or `1` or `12` etc.

(f) In case particular HTTP response received (too many requests) HTTP reply will not use rebounds but will retry after ever increasing delay

With this solution, the burden on counting number of clients calling on API as well as 'budget' allocation will be on integrator, however in my opinion this option have following advantages:

(d) Simpler implementation that can be done fast
(e) As no customer currently run and shout about this feature it's wiser to keep implementation efforts low and learn from the adoption

**tldr;** I would opt for going with simplified local rate control instead of proper centralized maestro-managed rate control manage and iterate from there.

**Comment by shulkaolka** [comment]
@zubairov @kirill-levitskiy as was discussed on planning, I implemented a solution with delay between requests (PR is [here](https://github.com/elasticio/rest-api-component/pull/133)). 
It is added 2 additional configuration fields:

> -  ``Delay`` - If you want to slow down requests to your API you can set delay value (in seconds) and the component will delay calling the next request after the previous request. 
> Time for the delay is calculated as `Delay`/ `Call Count` and shouldn't be more than 1140 seconds (19 minutes due to platform limitation). 
> The `Call Count` value by default is 1. If you want to use another value, please set the `Call Count` field. 
> 
> -  ``Call Count`` - the field should be used only in pair with `Delay`, default to 1.

Please, look at it. Is my implementation matches your expectation and can resolve the current issue?
